bbc/simorgh
# Investigation: force reload after period of inactivity

Resolves #simorgh-infrastructure/344

**Overall change:** Sets an inactivity period timer, which gets reset when the user navigates to a new route. If user doesn't navigate to a new route within 1 hour, their current window is automatically reloaded. This is to provide a degree of confidence that users who keep tabs open indefinitely will eventually be updated to the latest bundles.

**Code changes:**

- Adds a `reloadAfterInactivity` call to `client.js`, currently setting a 1 hr timeout (to be negotiated). I've set a minimum timeout period of 15 seconds as a security measure (minimise chances of accidentally DOSing ourselves).
- Creates an `onNewRoute` callback in `ClientApp`, which is triggered by changing routes (via clicking on Simorgh links). We use this callback to call `resetInteractivityTimer`, otherwise users who are actively reading an article will see the page unexpectedly reload because they've had an hour long reading session.
- Adds Jest tests for `reloadAfterInteractivity` (and its dependent, `objToMilliseconds`, which allows for a really nice readable API) and for `resetInteractivityTimer`. Also adds single unit tests to `ClientApp` to check that the `onNewRoute` callback is called when expected.

**Manual tests:**

- In addition to unit tests, I've manually set the timeout period to a shorter period and observed it working.
- I've manually observed clicking through to new routes resetting the timer.
- I've switched to new tabs and seen the article tab reload in the background when expected.
- I've minimised the window and then reopened the window and the timer continues to reload the page as expected.
- I've added a test that checks the page should reload even if an error is thrown by the page (it doesn't interfere with the `setTimeout` callback). Caveat being that the `setTimeout` and error are inside a try-catch. Provided we have some sort of error boundary (we should reopen #211), this should be sufficient.

**Considerations before merge**:

- Should test that the `setTimeout` callback is still called even if a runtime error occurs (_outside_ of a `try-catch`).
- Need to think through the timeout period. 1 hour was picked arbitrarily based on it being plenty of time to read an article and unlikely to disrupt the user experience.
- If the timeout period is long (say 24 hours) and the user has the window minified at the point that the timeout _would_ be called, do we risk the timeout never being called? Worth considering using something like the [Page Visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) to force a refresh if the user is coming back to a stale page after a period of inactivity.

---
